Backward through sparse_coo_tensor

aocsa opened the pull request:
Fixes #49683 

This PR  solves Backward through sparse_coo_tensor bug by implementing a `sparse_mask_helper` function for n-dimensional sparse tensor for CPU and CUDA which is used to reimplement `sparse_constructor_values_backward` function.  

This `sparse_mask` function was implemented before for  backward  sparse-sparse matmul. However,  the algorithm is little different  because in this case it should be applyable not only for matrices but for n-dimensional tensors. Thankfully it was not quite hard to extend and now both share the same code base.  

Note that  no new tests are required because now the backward for sparse-sparse matmul now uses the new `sparse_mask_helper`.

@ngimel, @mruberry - kindly review this.

aocsa committed:
first commit

aocsa committed:
Minor refactor. Closes gh-49683 

minor update

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2686,10 +2686,10 @@
     SparseCUDA: sparse_sparse_matmul_cuda
 
-- func: _sparse_matrix_mask_helper(Tensor t, Tensor mask_indices) -> Tensor
+- func: _sparse_mask_helper(Tensor t, Tensor mask_indices) -> Tensor
```
We should take this out of native_functions.yaml. @ngimel is going to write a DeveloperFAQ entry into how to support CPU and CUDA dispatch without a native_functions.yaml entry, @IvanYashchuk and @kurtamohler also did this recently, we think, and may be able to explain the process.

kurtamohler replied:
You can grep for `REGISTER_DISPATCH`, `DEFINE_DISPATCH`, and `DECLARE_DISPATCH` to see how to create a CPU/CUDA dispatch function. Usually (if not always), these functions are named `*_stub`. For instance, `eig_out` calls `eig_stub` to dispatch to the device-specific implementation.

aocsa replied:
I would like to do this in another PR. I tried to implement in that way in the native/sparse module and it doesn't work the the code is in this folder. There is a TODO, and I guess is the origin of the problem. https://github.com/pytorch/pytorch/blob/71ca600af93b4ad58ca6fc083fa5d42a56b43194/aten/src/ATen/native/DispatchStub.h#L37

mruberry commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -526,4 +526,57 @@ SparseTensor sparse_mask_cpu(const Tensor& t, const SparseTensor& mask) {
+
+    Note: every index exists in the sparse tensor `t`
+  */
+  int64_t r_nnz = mask_indices.size(1);
```
Does this assume that both tensors are coalesced?

aocsa replied:
This assumes that only input parameter `t` is coalesced. I added this note to the documentation

mruberry commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:
```diff
@@ -152,4 +198,91 @@ SparseTensor coalesce_sparse_cuda(const SparseTensor& self) {
+  auto t_nnz = t._nnz();
+
+  int curDevice = -1;
+  cudaGetDevice(&curDevice);
```
I think the default `getCurrentCUDAStream()` will give the stream on the current device?

aocsa replied:
yes you are right

mruberry commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:
```diff
@@ -152,4 +198,91 @@ SparseTensor coalesce_sparse_cuda(const SparseTensor& self) {
+      t_indices_pos.data_ptr<int64_t>());
+
+  // Step 4: Copy the Filtered `t._values()` using the matches at `t_indices_pos`
+  int64_t total_threads = r_nnz;
```
Style nit: just use r_nnz directly for readability

mruberry commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:
```diff
@@ -19,11 +19,57 @@
+  int64_t out_end = (i + 1) * values_stride0;
+  int64_t in_start = j * t_values_ti.strides[0];
+
+  if (has_match) {
```
Is this necessary to handle hybrid sparse-dense tensors?

aocsa replied:
This is necessary to copy just the values that has a match. The else part was  removed because the `r_values` are initialized as `Tensor r_values = at::zeros({vsize}, t_values.options());`

mruberry commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:
```diff
@@ -19,11 +19,57 @@
+    has_match = true;
+  }
+
+  int64_t values_stride0 = r_values_ti.strides[0];
```
How does this work with multidimensional dense tensors?

aocsa replied:
this this works with multidimensional sparse tensors like:

```python
def test_hybrid():
    indices = torch.LongTensor([[2, 3, 0], 
                                [7, 9, 0]])
    x = torch.tensor([  [[1,1], [2, 2]], 
                        [[2,2], [3, 3]],
                        [[4,4], [5, 5]],
                    ], dtype=torch.float32, requires_grad=True)
    print(x.shape)
    S = torch.sparse_coo_tensor(indices=indices, values=x, size=(10, 10, 2, 2), device='cuda')
    print(S)

    f = torch.sparse.sum(S) # calls to_dense(), allocating gigabytes of memory
    print("sum: ", f)
    f.backward()

    # print(S.grad)
    print(x.grad)
```
which are part of the test of `test_sparse_sum`

ngimel replied:
This would work only if values tensor is contiguous - is it guaranteed?

aocsa replied:
Yes it does, I added some TORCH_CHECKs at the beginning of the function.

mruberry commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -2866,15 +2866,8 @@ Tensor log1p_backward(const Tensor& grad, const Tensor& self) {
-  auto flattened_indices = at::sparse::flatten_indices(indices, full_size);
-  return flattened_dense_grad.index_select(0, flattened_indices);
+Tensor sparse_constructor_values_backward(const Tensor& sparse_grad_out, const Tensor& indices) {
+  return _sparse_mask_helper(sparse_grad_out, indices.contiguous());;
```
Nit: two semicolons at EoL

mruberry reviewed (commented):
@ngimel and I took a look and this looks like a good fix. We made a few comments inline. Can this also remove the previous functions:

https://github.com/pytorch/pytorch/blob/92f37ae263ac860e05ea7a4ff540052bb6934fed/aten/src/ATen/native/sparse/SparseMatMul.cpp#L208

and 

https://github.com/pytorch/pytorch/blob/92f37ae263ac860e05ea7a4ff540052bb6934fed/aten/src/ATen/native/sparse/cuda/SparseMatMul.cu#L767

?

Also, would you file a follow-up issue to add a test for this behavior? While that's a complicated issue, it would be interesting to think how we ensure this behavior is maintained in the future. @ngimel suggests running the operations under a profiler and verifying there is no .to_dense() call might work.

aocsa committed:
addressed review

aocsa commented:
Thanks @mruberry, I addressed review comments. 

> Also, would you file a follow-up issue to add a test for this behavior? While that's a complicated issue, it would be interesting to think how we ensure this behavior is maintained in the future. @ngimel suggests running the operations under a profiler and verifying there is no .to_dense() call might work.

Regarding the last comment, I would be interesting to use a profiler and verify that there is no `.to_dense()` call for other sparse operations. I didn't work before  with the [PyTorch profiler](https://pytorch.org/tutorials/recipes/recipes/profiler.html#use-profiler-to-analyze-memory-consumption), but I could look it and create new issues. cc @rgommers.

aocsa committed:
minor fix for contiguous data

aocsa committed:
fixes for edge cases

aocsa commented:
This PR was updated fixing some  test errors  reported for backward algorithm for cases when `numel() = 0` and `sparse_dim() = 0`. I think this PR is ready to be merged cc @mruberry.

ngimel commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -528,4 +528,63 @@ SparseTensor sparse_mask_cpu(const Tensor& t, const SparseTensor& mask) {
+      `t`             - coalesced sparse tensor input
+      `mask_indices`  - mask indices tensor
+
+    Note: every index exists in the sparse tensor `t`
```
Is it a requirement? I don't think it's guaranteed

aocsa replied:
Ohh my bad. You are right. it's not guaranteed

ngimel commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -528,4 +528,63 @@ SparseTensor sparse_mask_cpu(const Tensor& t, const SparseTensor& mask) {
+      int64_t index = flattened_mask_indices.data_ptr<int64_t>()[i];
+      auto iter = t_flatten_indices.find(index);
+      if (iter != t_flatten_indices.end()) {
+        assert(iter->second < t_nnz);
```
TORCH_INTERNAL_ASSERT, or remove it

ngimel commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:
```diff
@@ -152,4 +193,100 @@ SparseTensor coalesce_sparse_cuda(const SparseTensor& self) {
+    dim3 grid;
+    int curDevice = -1;
+    cudaGetDevice(&curDevice);
+    TORCH_CHECK(
```
Avoid using raw cuda APIs like cudaGetDevice, and getApplyGrid/Block (it will be deprecated), in this case you can use `at::cuda::getCurrentDeviceProperties()->maxGridSize[0],` to get max grid size, and ATenCeilDiv. Since your grid is 1D you don't even need comparing against maxGridSize (it's 2**31).

ngimel commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:
```diff
@@ -152,4 +193,100 @@ SparseTensor coalesce_sparse_cuda(const SparseTensor& self) {
+ 
+  if (t.sparse_dim() == 0) {
+    Tensor t_values_expand = t_values;
+    t_values_expand = t_values_expand.expand(vsize).clone(at::MemoryFormat::Contiguous);
```
you don't need `.clone()` here, just `.contiguous()` will do?

aocsa replied:
Yes, just contiguous

aocsa committed:
addressed review

aocsa commented:
Thanks @ngimel, I addressed review comments. cc @mruberry

ngimel commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -528,4 +528,59 @@ SparseTensor sparse_mask_cpu(const Tensor& t, const SparseTensor& mask) {
+      int64_t index = flattened_mask_indices.data_ptr<int64_t>()[i];
+      auto iter = t_flatten_indices.find(index);
+      if (iter != t_flatten_indices.end()) {
+        r_values[i] = t_v[ iter->second ];
```
if mask is uncoalesced then the same value will be written for each duplicate index, and after coalescing it will become t_v * n, where `n` is the number of duplicates of this index in an uncoalesced mask. I doubt this is desired semantics.

aocsa replied:
This function is used in two parts, `sparse_sparse_matmul_backward`  and `sparse_constructor_values_backward`. In the first case the mask is coalesced. In the second case the indices are the same as used in the function constructor `_sparse_coo_tensor_with_dims_and_tensors` and the `nnz` returned should be as same as the `indices`, so I think it's correct. Just in case I will add a note to the function doc.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -659,15 +659,15 @@ Tensor _sparse_addmm_sparse_backward(const Tensor& grad, const Tensor& sparse_,
 // Note that the `output` must have the same `indices` as the `mask` so we are using just a clone.
 // However, to get `values` we have to use specific helper function for CPU/CUDA and use the `mask` data to filter `values`
-// That's why we created this `_sparse_matrix_mask_helper` function.
+// That's why we created this `sparse_mask_helper` function.
```
nit missing `_`

aocsa committed:
minor update


minor update

aocsa commented:
This PR is updated addressing last comments cc @ngimel, @mruberry.
